Add scanpy/scrublet module

nictru opened the pull request:
These is the first of 13 scanpy modules. I am adding this as a preparation of the 1.0 release of nf-core/scdownstream

SPPearce commented on modules/nf-core/scanpy/scrublet/main.nf:
```diff
@@ -0,0 +1,47 @@
+    task.ext.when == null || task.ext.when
+
+    script:
+    prefix = task.ext.prefix ?: "${meta.id}"
```
Can we set a default prefix
```suggestion
    prefix = task.ext.prefix ?: "${meta.id}_scrubbed"
```
Whatever is most appropriate

nictru replied:
Yes of course - I can't remember the outcome in the maintainers discussion on this, but I am fine with adding it

nictru replied:
Done via [fed0b87](https://github.com/nf-core/modules/pull/8684/commits/fed0b877c8b85ed821218d9b98fc08a9dd01838e)

SPPearce commented on modules/nf-core/scanpy/scrublet/meta.yml:
```diff
@@ -0,0 +1,71 @@
+        type: map
+        description: |
+          Groovy Map containing sample information
+          e.g. [ id:'test', single_end:false ]
```
```suggestion
          e.g. [ id:'test' ]
```

SPPearce commented on modules/nf-core/scanpy/scrublet/meta.yml:
```diff
@@ -0,0 +1,71 @@
+          type: map
+          description: |
+            Groovy Map containing sample information
+            e.g. [ id:'test', single_end:false ]
```
```suggestion
            e.g. [ id:'test' ]
```

SPPearce commented on modules/nf-core/scanpy/scrublet/meta.yml:
```diff
@@ -0,0 +1,71 @@
+          type: map
+          description: |
+            Groovy Map containing sample information
+            e.g. [ id:'test', single_end:false ]
```
```suggestion
            e.g. [ id:'test' ]
```

SPPearce commented on nf-test.config:
```diff
@@ -19,5 +19,6 @@ config {
         load "nft-csv@0.1.0"
         load "nft-utils@0.0.3"
         load "nft-fastq@0.0.1"
+        load "nft-anndata@0.1.0"
```
Are you using this?

nictru replied:
Yes, I created nft-anndata for this purpose

SPPearce commented on modules/nf-core/scanpy/scrublet/main.nf:
```diff
@@ -0,0 +1,47 @@
+        error("Input and output names are the same, use \"task.ext.prefix\" to disambiguate!")
+    }
+    """
+    export MPLCONFIGDIR=./tmp/mpl
```
I'm guessing these are required to not give errors in the version information; if that is the case can you please add a comment to make sure nobody tries to remove them in the future.

SPPearce commented on modules/nf-core/scanpy/scrublet/tests/main.nf.test:
```diff
@@ -0,0 +1,69 @@
+    test("Should run without failures") {
+
+        when {
+            params {
+                outdir = "$outputDir"
+            }
```
```suggestion
```

SPPearce commented on modules/nf-core/scanpy/scrublet/tests/main.nf.test:
```diff
@@ -0,0 +1,69 @@
+nextflow_process {
+
+    name "Test Process SCANPY_SCRUBLET"
+    script "modules/nf-core/scanpy/scrublet/main.nf"
```
```suggestion
    script "../main.nf"
```

SPPearce commented on modules/nf-core/scanpy/scrublet/tests/main.nf.test:
```diff
@@ -0,0 +1,69 @@
+            assertAll(
+            { assert process.success },
+            { assert snapshot(process.out.versions).match() },
+            { assert "predicted_doublet" in anndata(process.out.h5ad[0][1]).obs.colnames }
```
Can we test the file names of the pkl files (assuming they are not stable).

SPPearce commented on modules/nf-core/scanpy/scrublet/tests/main.nf.test:
```diff
@@ -0,0 +1,69 @@
+        options '-stub'
+
+        when {
+            params {
+                outdir = "$outputDir"
+            }
```
```suggestion
```

SPPearce commented on modules/nf-core/scanpy/pca/main.nf:
```diff
@@ -0,0 +1,47 @@
+
+    output:
+    tuple val(meta), path("*.h5ad"), emit: h5ad
+    path "X_*.pkl"                 , emit: obsm
```
Should this not have the meta map?

nictru replied:
Removed PCA module from this PR, will fix in a different PR

SPPearce commented on modules/nf-core/scanpy/pca/meta.yml:
```diff
@@ -0,0 +1,65 @@
+  - - key_added:
+        type: string
+        description: |
+          Key under which to add the PCA results to adata.obsm
```
Can you give an example as to what this might be?

SPPearce commented on modules/nf-core/scanpy/pca/templates/pca.py:
```diff
@@ -0,0 +1,42 @@
+threadpool_limits(int("${task.cpus}"))
+sc.settings.n_jobs = int("${task.cpus}")
+
+
```
```suggestion
```

SPPearce commented on modules/nf-core/scanpy/pca/tests/main.nf.test:
```diff
@@ -0,0 +1,69 @@
+nextflow_process {
+
+    name "Test Process SCANPY_PCA"
+    script "modules/nf-core/scanpy/pca/main.nf"
```
```suggestion
    script "../main.nf"
```

SPPearce reviewed (commented):
Can you please make separate PRs for each module.

nictru commented:
I addressed the comments about scrublet and remobed the scanpy/pca module. I am now investigating the stability of the module a bit, maybe I can still get the hashes stable

nictru commented:
Seems to work

SPPearce commented:
> Can you please make separate PRs for each module.

Oh, sorry, now I read my previous comment I see it wasn't really what I meant to say (I blame it being 6am).
It should have read, "in the future"; this was fine having two modules, just easier if they are separated for reviewing.
